### PR TITLE
story #3033 Picking - Improve error messaging/logging when picking can't be validated

### PR DESCRIPTION
**Pull request by jrbirch99**
- A more detailed error message with information on the move lines in the "done" state are given
- Unit tests were added to test the error message
- Formatting

User-story: 3033

Signed-off-by: Jack Birch <jack.birch@unipart.com>

### CODE REVIEW

**Comment by jrbirch99** [comment]
I did formatting and changes in one commit... so the PR is a bit ugly. The only non-formatting change in stock_move_line.py are in `_prevent_write_after_done`.

**Comment by PeterAlabaster** [comment]
> I did formatting and changes in one commit... so the PR is a bit ugly. The only non-formatting change in stock_move_line.py are in `_prevent_write_after_done`.

nit: can we separate the black format into its own commit

File: addons/udes_stock/models/stock_move_line.py [hunk 1]
```diff
@@ -1110,14 +1115,35 @@ def _drop_off_criterion_summary(self):
         "qty_done",
     )
     def _prevent_write_after_done(self):
-        done_lines = self.filtered(lambda x: x.state == "done")
+        done_lines = self.filtered(lambda ml: ml.state == "done")
+
+        # When a moveline is updated through desktop only one "done" move line is passed into self even when there are multiple
+        # Get all the done move lines on the picking if there is only one "done" move line
+        if len(done_lines) == 1 and self.mapped("picking_id"):
+            done_lines = self.mapped("picking_id.move_line_ids").filtered(
```
Comment (PeterC10) [inline addons/udes_stock/models/stock_move_line.py:1123]: Suggestion: This mapped could be pretty expensive on big databases, and this constraint will be checked fairly often.

Can we try using `search` or `search_read` here?

File: addons/udes_stock/models/stock_move_line.py [hunk 2]
```diff
@@ -1110,14 +1115,35 @@ def _drop_off_criterion_summary(self):
...
+                _logger.error(
+                    f"Move Line id: {line.id} in state 'done' cannot be written to. Product id: {line.product_id.id}, Quantity: {line.qty_done}, Destination Location id: {line.location_dest_id.id}, Lot/Serial id: {line.lot_id if line.lot_id else False}, Package id: {line.result_package_id.id if line.result_package_id else False}"
+                )
+                move_line_strings.append(
+                    f"Product: {line.product_id.name}, Quantity: {line.qty_done}, Location: {line.location_dest_id.name}, Lot/Serial Number: {line.lot_name if line.lot_name else False}, Package Name: {line.result_package_id.name if line.result_package_id else False}"
+                )
+
+            raise ValidationError(
+                _("Cannot update move lines that are already 'done': \n%s")
+                % ("\n".join(move_line_strings))
```
Comment (PeterC10) [inline addons/udes_stock/models/stock_move_line.py:1141]: Can we include the Picking name?

File: addons/udes_stock/models/stock_move_line.py [hunk 3]
```diff
@@ -1110,14 +1115,35 @@ def _drop_off_criterion_summary(self):
...
+                lambda ml: ml.state == "done"
+            )
+
         if done_lines:
-            raise ValidationError(_("Cannot update move lines that are already 'done'."))
+            # For each done line
+            # Create error message
+            move_line_strings = list()
+            for line in done_lines:
+                _logger.error(
```
Comment (PeterC10) [inline addons/udes_stock/models/stock_move_line.py:1132]: Picking name and/or ID would be useful here too.

File: addons/udes_stock/models/stock_move_line.py [hunk 4]
```diff
@@ -1110,14 +1115,35 @@ def _drop_off_criterion_summary(self):
...
+                    f"Move Line id: {line.id} in state 'done' cannot be written to. Product id: {line.product_id.id}, Quantity: {line.qty_done}, Destination Location id: {line.location_dest_id.id}, Lot/Serial id: {line.lot_id if line.lot_id else False}, Package id: {line.result_package_id.id if line.result_package_id else False}"
```
Comment (PeterC10) [inline addons/udes_stock/models/stock_move_line.py:1133]: Nit: For logged statements we should use `%` for strings.

File: addons/udes_stock/models/stock_move_line.py (hunk 1 above)
Comment (jrbirch99) [reply addons/udes_stock/models/stock_move_line.py:1123]: I had thought about this, but I didn't think there would be much of a performance dip as the mapped will only ever be called if self contains `done_lines` which might be quite rare.   

Happy to change it though

File: addons/udes_stock/models/stock_move_line.py (hunk 3 above)
Comment (jrbirch99) [reply addons/udes_stock/models/stock_move_line.py:1131]: changed

File: addons/udes_stock/models/stock_move_line.py (hunk 4 above)
Comment (jrbirch99) [reply addons/udes_stock/models/stock_move_line.py]: changed

File: addons/udes_stock/models/stock_move_line.py (hunk 2 above)
Comment (jrbirch99) [reply addons/udes_stock/models/stock_move_line.py:1147]: changed

**Review by PeterAlabaster — APPROVED** [review]

File: addons/udes_stock/models/stock_move_line.py [hunk 5]
```diff
@@ -1120,20 +1120,26 @@ def _prevent_write_after_done(self):
         # When a moveline is updated through desktop only one "done" move line is passed into self even when there are multiple
         # Get all the done move lines on the picking if there is only one "done" move line
         if len(done_lines) == 1 and self.mapped("picking_id"):
-            done_lines = self.mapped("picking_id.move_line_ids").filtered(
-                lambda ml: ml.state == "done"
+            done_lines = self.search(
+                [("picking_id", "in", self.mapped("picking_id.id")), ("state", "=", "done")]
```
Comment (PeterC10) [inline addons/udes_stock/models/stock_move_line.py:1124]: Couple of minor suggestions:

```
if len(done_lines) == 1:
    picking_ids = self.mapped("picking_id.id")
    if picking_ids:
        done_lines = self.search([("picking_id", "in", self.mapped("picking_id.id")), ("state", "=", "done")], order="id")
```

This avoids us doing `mapped` twice. And `order="id"` means we avoid using the default `_order` on the move line model, as the order of records doesn't matter here.

File: addons/udes_stock/models/stock_move_line.py [hunk 6]
```diff
@@ -1119,10 +1119,12 @@ def _prevent_write_after_done(self):
...
         # Get all the done move lines on the picking if there is only one "done" move line
-        if len(done_lines) == 1 and self.mapped("picking_id"):
-            done_lines = self.search(
-                [("picking_id", "in", self.mapped("picking_id.id")), ("state", "=", "done")]
-            )
+        if len(done_lines) == 1:
+            picking_ids = self.mapped("picking_id")
+            if picking_ids:
+                done_lines = self.search(
+                    [("picking_id", "in", self.mapped("picking_id.id")), ("state", "=", "done")], order="id"
```
Comment (PeterC10) [inline addons/udes_stock/models/stock_move_line.py:1126]: Mistake on my comment sorry, should be `("picking_id", "in", picking_ids)` - as we already have the picking_ids from line 1123.

**Review by PeterC10 — APPROVED** [review]

**Pull request merged**